src/NSrtm.Core/Pgm/PgmElevationProvider.cs

```diff
@@ -0,0 +1,119 @@
+
+        public PgmElevationProvider([NotNull] IPgmGeoidUndulationGrid discreteSurface)
+        {
+            if (discreteSurface == null) throw new ArgumentNullException("discreteSurface");
+            _discreteSurface = discreteSurface;
+        }
+
+        private PgmDataDescription dataDescription { get { return _discreteSurface.PgmParameters; } }
+
+        private BivariatePolynomial getInterpolationForCellSurface(PgmCellCoords pgmCellCoords)
+        {
+            var nodesCoordinates = findCellAndSurroundingNodesCoords(pgmCellCoords,
+                                                                     dataDescription.LatitudeIncrementDegrees,
+                                                                     dataDescription.LongitudeIncrementDegrees);
+            var nodesUndulations = nodesCoordinates.Select(coor => _discreteSurface.GetUndulation(coor.Lat, coor.Lon));
+            var formattedUndulations = nodesUndulations.Select((c, i) => new {Index = i, value = c})
+                                                       .GroupBy(p => p.Index / 4)
+                                                       .Select(c => c.Select(v => v.value)
+                                                                     .ToList())
+                                                       .ToList();
+            return BicubicCalculator.GetSpline(values: formattedUndulations, step: dataDescription.LatitudeIncrementDegrees);
+        }
+
+        private static PgmCellCoords normalizeCoords(double lat, double lon)
+        {
+            var normalizedLongitude = lon;
+            if (lat > 90 || lat < -90)
+            {
+                normalizedLongitude += 180;
+            }
+            normalizedLongitude = (360 + (normalizedLongitude % 360)) % 360;
+            var absoluteLat = Math.Abs(Math.Abs(lat % 180) - 90);
+            var normalizedLatitude = lat > 0 ? 90 - absoluteLat : absoluteLat - 90;
+            return new PgmCellCoords(normalizedLatitude, normalizedLongitude);
+        }
+
+        /// <summary>
+        ///     Gets elevation of the geoid above the ellipsoid.
+        /// </summary>
+        /// <param name="latitude">Latitude in degrees in WGS84 datum</param>
+        /// <param name="longitude">Longitude in degrees in WGS84 datum</param>
+        /// <returns></returns>
+        public double GetElevation(double latitude, double longitude)
+        {
+            var longitudeInEgmDatum = longitude + 180;
+            var mainNode = PgmCellCoords.ForCoordinatesUsingDescription(latitude, longitudeInEgmDatum, dataDescription);
+            var interpolatedCell = _continuousSurface.GetOrAdd(mainNode, getInterpolationForCellSurface);
+            return interpolatedCell.Evaluate(latitude, longitudeInEgmDatum);
+        }
+
+        public Task<double> GetElevationAsync(double latitude, double longitude)
+        {
+            return Task.FromResult(GetElevation(latitude, longitude));
```

Comment:
PgmElevationProvider get grid in constructor. Get elevation only create bivariatepolygonal if necessary and take elevation.